### PR TITLE
feat(vmclass): add validation for matching virtual machine sizing policies upon virtual machine class change

### PR DESCRIPTION
**Pull request by eofff**
## Description
A validation has been added to ensure virtual machines match the sizing policies when their virtual machine class is changed. The result of this matching check is also added to the `conditions` field of the virtual machine.

## What is the expected result?
The matching status of the virtual machine with the sizing policies should be reflected in the corresponding `condition`.

## Checklist
- [ ] The code is covered by unit tests.
- [ ] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

### CODE REVIEW

File: images/virtualization-artifact/pkg/controller/vm/internal/size_policy.go [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+	if update := addAllUnknown(changed, agentConditions...); update {
+		return reconcile.Result{Requeue: true}, nil
+	}
+
+	if isDeletion(current) {
+		return reconcile.Result{}, nil
+	}
+
+	//nolint:staticcheck
+	mgr := conditions.NewManager(changed.Status.Conditions)
```
Comment (yaroslavborbat) [inline images/virtualization-artifact/pkg/controller/vm/internal/size_policy.go:60]: don't use manager

File: images/virtualization-artifact/pkg/controller/vm/internal/size_policy.go [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+}
+
+func (h *SizePolicyHandler) Handle(ctx context.Context, s state.VirtualMachineState) (reconcile.Result, error) {
+	if s.VirtualMachine().IsEmpty() {
+		return reconcile.Result{}, nil
+	}
+	current := s.VirtualMachine().Current()
+	changed := s.VirtualMachine().Changed()
+
+	if update := addAllUnknown(changed, agentConditions...); update {
```
Comment (yaroslavborbat) [inline images/virtualization-artifact/pkg/controller/vm/internal/size_policy.go:51]: agentConditions?

File: images/virtualization-artifact/pkg/controller/vm/internal/size_policy.go [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+			Reason(vmcondition.ReasonSizingPolicyMatched).
+			Status(metav1.ConditionTrue)
+	} else {
+		cb.Message(err.Error()).
+			Reason(vmcondition.ReasonSizingPolicyNotMatched).
+			Status(metav1.ConditionFalse)
+	}
+
+	mgr.Update(cb.Condition())
+	changed.Status.Conditions = mgr.Generate()
```
Comment (yaroslavborbat) [inline images/virtualization-artifact/pkg/controller/vm/internal/size_policy.go:76]:
```suggestion
	conditions.SetCondition(cb, &changed.Status.Conditions)
```

File: images/virtualization-artifact/pkg/controller/vm/vm_reconciler.go [hunk 4]
```diff
@@ -237,6 +237,54 @@ func (r *Reconciler) SetupController(_ context.Context, mgr manager.Manager, ctr
...
+			}
+
+			c := mgr.GetClient()
+			vms := &virtv2.VirtualMachineList{}
+			err := c.List(ctx, vms, client.MatchingFields{
+				indexer.IndexFieldVMByClass: class.GetName(),
+			})
+			if err != nil {
+				log := logger.FromContext(ctx)
+				log.Error(err.Error())
```
Comment (yaroslavborbat) [inline images/virtualization-artifact/pkg/controller/vm/vm_reconciler.go:257]: add more context and use logger.SlogErr attr

File: images/virtualization-artifact/pkg/controller/vm/vm_reconciler.go [hunk 5]
```diff
@@ -237,6 +237,54 @@ func (r *Reconciler) SetupController(_ context.Context, mgr manager.Manager, ctr
...
+		predicate.Funcs{
+			CreateFunc: func(e event.CreateEvent) bool { return false },
+			DeleteFunc: func(e event.DeleteEvent) bool { return false },
+			UpdateFunc: func(e event.UpdateEvent) bool {
+				oldVMC, oldOk := e.ObjectOld.(*virtv2.VirtualMachineClass)
+				newVMC, newOk := e.ObjectNew.(*virtv2.VirtualMachineClass)
+				if !oldOk || !newOk {
+					return false
+				}
+				return !reflect.DeepEqual(oldVMC, newVMC)
```
Comment (yaroslavborbat) [inline images/virtualization-artifact/pkg/controller/vm/vm_reconciler.go:281]: compare resource spec. Not all

**Review by yaroslavborbat — CHANGES_REQUESTED** [review]

File: images/virtualization-artifact/pkg/controller/vm/internal/size_policy.go (hunk 1 above)
Comment (yaroslavborbat) [reply images/virtualization-artifact/pkg/controller/vm/internal/size_policy.go:60]: delete it

**Review by yaroslavborbat — APPROVED** [review]

**Review by yaroslavborbat — DISMISSED** [review]

File: images/virtualization-artifact/pkg/controller/vm/internal/size_policy.go [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+	if isDeletion(current) {
+		return reconcile.Result{}, nil
+	}
+
+	cb := conditions.NewConditionBuilder(vmcondition.TypeSizingPolicyMatched).
+		Generation(current.GetGeneration())
+
+	err := h.service.CheckVMMatchedSizePolicy(ctx, changed)
+	if err == nil {
+		cb.Message("").
```
Comment (Isteb4k) [inline images/virtualization-artifact/pkg/controller/vm/internal/size_policy.go:64]: No need to call Message(""), empty string is already there

File: images/virtualization-artifact/pkg/controller/vm/internal/size_policy.go [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+			Reason(vmcondition.ReasonSizingPolicyMatched).
+			Status(metav1.ConditionTrue)
+	} else {
+		cb.Message(err.Error()).
```
Comment (Isteb4k) [inline images/virtualization-artifact/pkg/controller/vm/internal/size_policy.go:68]: The first letter in the condition must be uppercase, and the message must end with a dot.

**Review by Isteb4k — CHANGES_REQUESTED** [review]

File: images/virtualization-artifact/pkg/controller/vm/internal/watcher/vmclass_watcher.go [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+					NamespacedName: types.NamespacedName{
+						Name:      vm.Name,
+						Namespace: vm.Namespace,
+					},
+				}
+			}
+			return requests
+		}),
+		predicate.Funcs{
+			CreateFunc: func(e event.CreateEvent) bool { return false },
```
Comment (Isteb4k) [inline images/virtualization-artifact/pkg/controller/vm/internal/watcher/vmclass_watcher.go:79]: Are you sure there's no need to watch create and update events? What if a vmclass was missed but has now been created?

**Review by Isteb4k — CHANGES_REQUESTED** [review]

File: images/virtualization-artifact/pkg/controller/vm/internal/validators/sizing_policy_validator.go [hunk 9]
```diff
@@ -18,28 +18,78 @@ package validators
...
+			log := logger.FromContext(ctx)
+			log.Error(
+				"An error occurred while retrieving the VM class.",
+				logger.SlogErr(err),
+			)
+			return nil, err
+		}
+	}
+
+	return nil, v.service.CheckVMMatchedSizePolicy(vm, vmClass)
```
Comment (yaroslavborbat) [inline images/virtualization-artifact/pkg/controller/vm/internal/validators/sizing_policy_validator.go:68]: don't repeat yourself

File: images/virtualization-artifact/pkg/controller/vm/internal/size_policy.go [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+	changed := s.VirtualMachine().Changed()
+
+	addAllUnknown(changed, vmcondition.TypeSizingPolicyMatched)
+
+	if isDeletion(current) {
+		return reconcile.Result{}, nil
+	}
+
+	vmClass, err := s.Class(ctx)
+	if err != nil && !errors.IsNotFound(err) {
```
Comment (yaroslavborbat) [inline images/virtualization-artifact/pkg/controller/vm/internal/size_policy.go:61]: helper.FetchObject is used by the s.Class method. No need to handle "not found" error

**Review by yaroslavborbat — CHANGES_REQUESTED** [review]

File: images/virtualization-artifact/pkg/controller/vm/internal/size_policy.go [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+func (h *SizePolicyHandler) Handle(ctx context.Context, s state.VirtualMachineState) (reconcile.Result, error) {
+	if s.VirtualMachine().IsEmpty() {
+		return reconcile.Result{}, nil
+	}
+	current := s.VirtualMachine().Current()
+	changed := s.VirtualMachine().Changed()
+
+	addAllUnknown(changed, vmcondition.TypeSizingPolicyMatched)
+
+	if isDeletion(current) {
```
Comment (Isteb4k) [inline images/virtualization-artifact/pkg/controller/vm/internal/size_policy.go:56]: The condition is not set here when you return

**Review by Isteb4k — CHANGES_REQUESTED** [review]

**Review by Isteb4k — APPROVED** [review]

File: images/virtualization-artifact/pkg/controller/vm/internal/size_policy.go [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+		return reconcile.Result{}, nil
+	}
+
+	vmClass, err := s.Class(ctx)
+	if err != nil {
+		log := logger.FromContext(ctx)
+		log.Error(
+			"An error occurred while retrieving the VM class.",
+			logger.SlogErr(err),
+		)
```
Comment (yaroslavborbat) [inline images/virtualization-artifact/pkg/controller/vm/internal/size_policy.go:70]:
```suggestion
		return reconcile.Result{}, err
```

**Review by yaroslavborbat — CHANGES_REQUESTED** [review]

**Review by yaroslavborbat — APPROVED** [review]

**Review by yaroslavborbat — APPROVED** [review]

**Pull request merged**